core/src/main/java/org/elasticsearch/script/ScriptType.java

```diff
@@ -24,66 +24,117 @@
+     * from disk.  They are turned on by default as they should always be safe to execute.
+     */
+    FILE   ( 2 , "file"   , new ParseField("file")         , true  );
 
+    /**
+     * Reads an int from the input stream and converts it to a {@link ScriptType}.
+     * @return The ScriptType read from the stream. Throws an {@link IllegalStateException} if no ScriptType is found based on the id.
+     */
     public static ScriptType readFrom(StreamInput in) throws IOException {
-        int scriptTypeVal = in.readVInt();
-        for (ScriptType type : values()) {
-            if (type.val == scriptTypeVal) {
-                return type;
-            }
-        }
-        throw new IllegalArgumentException("Unexpected value read for ScriptType got [" + scriptTypeVal + "] expected one of ["
-                + INLINE.val + "," + FILE.val + "," + STORED.val + "]");
-    }
+        int id = in.readVInt();
 
-    public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOException{
-        if (scriptType != null) {
-            out.writeVInt(scriptType.val);
+        if (FILE.id == id) {
+            return FILE;
+        } else if (STORED.id == id) {
+            return STORED;
+        } else if (INLINE.id == id) {
+            return INLINE;
         } else {
-            out.writeVInt(INLINE.val); //Default to inline
+            throw new IllegalStateException("Error reading ScriptType id [" + id + "] from stream, expected one of [" +
+                FILE.id + " [" + FILE.name + "], " + STORED.id + " [" + STORED.name + "], " + INLINE.id + " [" + INLINE.name + "]]");
         }
     }
 
-    ScriptType(int val, String name, String scriptType, boolean defaultScriptEnabled) {
-        this.val = val;
-        this.parseField = new ParseField(name);
-        this.scriptType = scriptType;
-        this.defaultScriptEnabled = defaultScriptEnabled;
+    /**
+     * Writes an int to the output stream based on the id of the {@link ScriptType}.
+     * @param type The {@link ScriptType}.  Must not be {@code null}.
+     */
+    public static void writeTo(ScriptType type, StreamOutput out) throws IOException {
+        Objects.requireNonNull(type);
+        out.writeVInt(type.id);
     }
 
-    public ParseField getParseField() {
-        return parseField;
+    private final int id;
+    private final String name;
+    private final ParseField parseField;
+    private final boolean defaultEnabled;
+
+    /**
+     * Standard constructor.
+     * @param id A unique identifier for a type that can be read/written to a stream.
+     * @param name A unique name for a type.
+     * @param parseField Specifies the name used to parse input from queries.
+     * @param defaultEnabled Whether or not an {@link ScriptType} can be run by default.
+     */
+    ScriptType(int id, String name, ParseField parseField, boolean defaultEnabled) {
+        this.id = id;
+        this.name = name;
```

Comment:
The name is always the `name` in the `ParseField` so maybe we should just get it from there?